### PR TITLE
fixes and features for the pg_hba module

### LINKED ISSUE #776: The `keep_comments_at_rules` option of `postgresql_pg_hba` breaks idempotency
##### SUMMARY
if `keep_comments_at_rules` is `false` (which is the default) this breaks idempotency

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
`postgresql_pg_hba`

##### ANSIBLE VERSION
```
% ansible --version
ansible [core 2.17.0]
  config file = None
  configured module search path = ['/home/tommy/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/tommy/venvs/ansible/lib/python3.11/site-packages/ansible
  ansible collection location = /home/tommy/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/tommy/venvs/ansible/bin/ansible
  python version = 3.11.9 (main, Jul 11 2024, 14:01:32) [GCC 9.4.0] (/home/tommy/venvs/ansible/bin/python3.11)
  jinja version = 3.1.2
  libyaml = True
```

##### COLLECTION VERSION
All versions of the collections since `1.5.0`

##### STEPS TO REPRODUCE
Run this playbook twice:
```
- hosts: all
  tasks:

    - name: Create a rule with the comment 'comment1'
      postgresql_pg_hba:
        contype: host
        dest: /tmp/pg_hba2.conf
        create: true
        method: md5
        address: "2001:db8::1/128"
        state: present
        comment: "comment1"
        keep_comments_at_rules: false

    - shell: cat /tmp/pg_hba2.conf
      register: out1
    - debug:
        var: out1

    - name: Create a rule with the comment 'comment2'
      postgresql_pg_hba:
        contype: host
        dest: /tmp/pg_hba2.conf
        method: md5
        address: "2001:db8::2/128"
        state: present
        comment: "comment2"
        keep_comments_at_rules: false

    - shell: cat /tmp/pg_hba2.conf
      register: out1
    - debug:
        var: out1
```

```
% ansible-playbook -i inv.yml pg-hba-test.yml

PLAY [all] ***********************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************
ok: [local]

TASK [Create a rule with the comment 'comment1'] *********************************************************************************************************************************************
changed: [local]

TASK [shell] *********************************************************************************************************************************************************************************
changed: [local]

TASK [debug] *********************************************************************************************************************************************************************************
ok: [local] => {
    "out1": {
        "changed": true,
...
        "stdout": "\nhost\tall\tall\t2001:db8::1/128\tmd5\t#comment1",
        "stdout_lines": [
            "",
            "host\tall\tall\t2001:db8::1/128\tmd5\t#comment1"
        ]
    }
}

TASK [Create a rule with the comment 'comment2'] *********************************************************************************************************************************************
changed: [local]

TASK [shell] *********************************************************************************************************************************************************************************
changed: [local]

TASK [debug] *********************************************************************************************************************************************************************************
ok: [local] => {
    "out1": {
        "changed": true,
...
        "stdout": "#comment1\nhost\tall\tall\t2001:db8::1/128\tmd5\nhost\tall\tall\t2001:db8::2/128\tmd5\t#comment2",
        "stdout_lines": [
            "#comment1",
            "host\tall\tall\t2001:db8::1/128\tmd5",
            "host\tall\tall\t2001:db8::2/128\tmd5\t#comment2"
        ]
    }
}

PLAY RECAP ***********************************************************************************************************************************************************************************
local                      : ok=7    changed=4    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

tommy@passau ~/test % ansible-playbook -i inv.yml pg-hba-test.yml

PLAY [all] ***********************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************
ok: [local]

TASK [Create a rule with the comment 'comment1'] *********************************************************************************************************************************************
changed: [local]

TASK [shell] *********************************************************************************************************************************************************************************
changed: [local]

TASK [debug] *********************************************************************************************************************************************************************************
ok: [local] => {
    "out1": {
        "changed": true,
...
        "stdout": "#comment1\n#comment2\nhost\tall\tall\t2001:db8::1/128\tmd5\t#comment1\nhost\tall\tall\t2001:db8::2/128\tmd5",
        "stdout_lines": [
            "#comment1",
            "#comment2",
            "host\tall\tall\t2001:db8::1/128\tmd5\t#comment1",
            "host\tall\tall\t2001:db8::2/128\tmd5"
        ]
    }
}

TASK [Create a rule with the comment 'comment2'] *********************************************************************************************************************************************
changed: [local]

TASK [shell] *********************************************************************************************************************************************************************************
changed: [local]

TASK [debug] *********************************************************************************************************************************************************************************
ok: [local] => {
    "out1": {
        "changed": true,
...
        "stdout": "#comment1\n#comment2\n#comment1\nhost\tall\tall\t2001:db8::1/128\tmd5\nhost\tall\tall\t2001:db8::2/128\tmd5\t#comment2",
        "stdout_lines": [
            "#comment1",
            "#comment2",
            "#comment1",
            "host\tall\tall\t2001:db8::1/128\tmd5",
            "host\tall\tall\t2001:db8::2/128\tmd5\t#comment2"
        ]
    }
}

PLAY RECAP ***********************************************************************************************************************************************************************************
local                      : ok=7    changed=4    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

##### EXPECTED RESULTS
As shown above the same comments are added and moved over and over again

##### ACTUAL RESULTS
The comments don't get added several times

##### SUGGESTED FIX
Completely remove the `keep_comments_at_rules` option and add a way to add a comment to the top of the file.

### LINKED ISSUE #795: Support for include_dir
##### SUMMARY
This collection needs to support the new directive include_dir and a couple more that are present since postgres 16 (or at least not fail when they are present), in this blog they explain the new directives [here](https://opensource-db.com/unleashing-pg16-new-features-pghba-features/)

##### ISSUE TYPE
With current version 3.10 there's an error when the file  pg_hba.conf has any of these:
```
include             file
include_if_exists   file
include_dir         directory
```

##### COMPONENT NAME

community.postgresql.postgresql_pg_hba

##### ANSIBLE VERSION

```
ansible [core 2.16.3]
  config file = None
  configured module search path = ['/home/jjuvan/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /home/jjuvan/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.12.7 (main, Jan 17 2025, 16:55:27) [GCC 14.2.0] (/usr/bin/python3)
  jinja version = 3.1.3
  libyaml = True

```

##### COLLECTION VERSION

```
# /home/jjuvan/.ansible/collections/ansible_collections
Collection           Version
-------------------- -------
community.postgresql 3.10.0 

```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

Ubuntu 24.10

##### STEPS TO REPRODUCE

Any snipped like this
```
    community.postgresql.postgresql_pg_hba:
      dest: /etc/postgresql/17/main/pg_hba.conf
      contype: host
      users: "test"
      source: "192.168.1.11/32"
      comment: "test"
      databases: "test"
      method: md5
      keep_comments_at_rules: true
      backup: true
```

But with one of the new directives in the pg_hba.conf file

##### EXPECTED RESULTS

New entry being created on the file

##### ACTUAL RESULTS

Rule not being applied and a message about missing colums (that is normal in this case)

> "module_stdout": "Traceback (most recent call last):\r\n  File \"/home/bassols/.ansible/tmp/ansible-tmp-1737460694.0555177-231767-112376591644504/AnsiballZ_postgresql_pg_hba.py\", line 107, in <module>\r\n    _ansiballz_main()\r\n  File \"/home/bassols/.ansible/tmp/ansible-tmp-1737460694.0555177-231767-112376591644504/AnsiballZ_postgresql_pg_hba.py\", line 99, in _ansiballz_main\r\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\r\n  File \"/home/bassols/.ansible/tmp/ansible-tmp-1737460694.0555177-231767-112376591644504/AnsiballZ_postgresql_pg_hba.py\", line 47, in invoke_module\r\n    runpy.run_module(mod_name='ansible_collections.community.postgresql.plugins.modules.postgresql_pg_hba', init_globals=dict(_module_fqn='ansible_collections.community.postgresql.plugins.modules.postgresql_pg_hba', _modlib_path=modlib_path),\r\n  File \"/usr/lib/python3.10/runpy.py\", line 224, in run_module\r\n    return _run_module_code(code, init_globals, run_name, mod_spec)\r\n  File \"/usr/lib/python3.10/runpy.py\", line 96, in _run_module_code\r\n    _run_code(code, mod_globals, init_globals,\r\n  File \"/usr/lib/python3.10/runpy.py\", line 86, in _run_code\r\n    exec(code, run_globals)\r\n  File \"/tmp/ansible_community.postgresql.postgresql_pg_hba_payload_6g3voh9x/ansible_community.postgresql.postgresql_pg_hba_payload.zip/ansible_collections/community/postgresql/plugins/modules/postgresql_pg_hba.py\", line 1358, in <module>\r\n  File \"/tmp/ansible_community.postgresql.postgresql_pg_hba_payload_6g3voh9x/ansible_community.postgresql.postgresql_pg_hba_payload.zip/ansible_collections/community/postgresql/plugins/modules/postgresql_pg_hba.py\", line 1259, in main\r\n  File \"/tmp/ansible_community.postgresql.postgresql_pg_hba_payload_6g3voh9x/ansible_community.postgresql.postgresql_pg_hba_payload.zip/ansible_collections/community/postgresql/plugins/modules/postgresql_pg_hba.py\", line 495, in __init__\r\n  File \"/tmp/ansible_community.postgresql.postgresql_pg_hba_payload_6g3voh9x/ansible_community.postgresql.postgresql_pg_hba_payload.zip/ansible_collections/community/postgresql/plugins/modules/postgresql_pg_hba.py\", line 526, in read\r\n  File \"/tmp/ansible_community.postgresql.postgresql_pg_hba_payload_6g3voh9x/ansible_community.postgresql.postgresql_pg_hba_payload.zip/ansible_collections/community/postgresql/plugins/modules/postgresql_pg_hba.py\", line 572, in add_rule\r\n  File \"/tmp/ansible_community.postgresql.postgresql_pg_hba_payload_6g3voh9x/ansible_community.postgresql.postgresql_pg_hba_payload.zip/ansible_collections/community/postgresql/plugins/modules/postgresql_pg_hba.py\", line 993, in key\r\n  File \"/tmp/ansible_community.postgresql.postgresql_pg_hba_payload_6g3voh9x/ansible_community.postgresql.postgresql_pg_hba_payload.zip/ansible_collections/community/postgresql/plugins/modules/postgresql_pg_hba.py\", line 741, in source\r\n  File \"/usr/lib/python3.10/ipaddress.py\", line 83, in ip_network\r\n    raise ValueError(f'{address!r} does not appear to be an IPv4 or IPv6 network')\r\nValueError: 'None/None' does not appear to be an IPv4 or IPv6 network\r\n",

### PR DESCRIPTION
**Pull request by toydarian**
##### SUMMARY

 - Adding myself to the CODEOWNERS file for the parts related to `postgresql_pg_hba`
 - Deprecates `keep_comments_at_rules` as discussed in #776 
 - Adds parameter `sort_rules` that allows to disable sorting of rules. The default is `true` to keep to previous behavior
 - Allows a file to contain `include` directives and will keep them at the end of the file. Does not allow to add such directives (#795)
 - Adds the rendered string that is written to the file as an output

Fixes #776 
Fixes #795 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
`postgresql_pg_hba`

##### ADDITIONAL INFORMATION

- Removes the old `PgHba` class and moves the code to different stand-alone methods
- Adds tests for several edge-cases and new features
- modification in the tests is to account for the changes regarding `keep_comments_at_rules`
- adds required fetch in tests

### CODE REVIEW

**Comment by Andersson007** [comment]
@toydarian thanks for a lot of good changes! Could you please split the PR into one per changelog entry?

**Comment by Andersson007** [comment]
> I can try. It's actually hard, as a lot of the changes are closely tied in together and I want to avoid writing code that will be removed in the next PR.

sure, any smaller PRs would be much appreciated, thanks

**Review by hunleyd — CHANGES_REQUESTED** [review]
marking as 'changes requested' so my Slack bot stops reminding me about this. once it is out of draft, please re-request me @toydarian . thx!

**Comment by Andersson007** [comment]
Just a reminder to change it from draft to ready for review one it's ready, thanks

**Comment by toydarian** [comment]
@Andersson007 for some reason, when I use the methods from `module`, it fails on SELinux stuff in RHEL in the [CI](https://dev.azure.com/ansible/community.postgresql/_build/results?buildId=136127&view=logs&j=0e7ff078-b0f2-52bd-ae5e-08bbb58fe076&t=88b04eb7-73a8-570e-16db-fd23ef705b06). I don't have one of those at my disposal and the container is also not available for local testing. So I can't really test if it is only in the CI or what is actually going on.  
To make the CI work, I just do all the file-stuff myself, but I would like to do it properly.

So I'm pondering if I should try to get it to work or just make sure the perms are properly set and move the SELinux stuff to a new issue.

**Comment by Andersson007** [comment]
@toydarian sorry for the delayed response: the conference. To me it looks like the checks are OK now? Anyway, if there are any issues with a specific distro we can't fix on our side, you could just exclude it using `when:` in the test target

**Comment by Andersson007** [comment]
cc @betanummeric @sebasmannem

File: plugins/modules/postgresql_pg_hba.py [hunk 1]
```diff
@@ -1227,13 +1276,16 @@ def main():
         dest=dict(type='path', required=True),
         method=dict(type='str', default='md5', choices=PG_HBA_METHODS),
         netmask=dict(type='str'),
+        # TODO this can probably be changed to dict without breaking
         options=dict(type='str'),
+        # DEPRECATED, does nothing
```
Comment (Andersson007) [inline plugins/modules/postgresql_pg_hba.py:1281]:
```suggestion
        # DEPRECATED, does nothing
```
should it be deprecated deprecated? please see https://docs.ansible.com/ansible/latest/dev_guide/developing_program_flow_modules.html#argument-spec , there are corresponding fields similar to as we used when deprecating aliases

File: plugins/modules/postgresql_pg_hba.py [hunk 2]
```diff
@@ -1307,50 +1367,80 @@ def main():
...
                         # use module defaults
                         rule[key] = argument_spec[key].get('default', None)
-            new_rules.append(rule)
-        rules = new_rules
 
-    for rule in rules:
-        if rule.get('contype', None) is None:
-            continue
+            new_rules.append(rule)
+    # TODO
```
Comment (Andersson007) [inline plugins/modules/postgresql_pg_hba.py:1372]:
```suggestion
    # TODO
```
does it make sense to move these lines to a separate function for readability?

File: plugins/modules/postgresql_pg_hba.py [hunk 3]
```diff
@@ -1274,11 +1330,16 @@ def main():
         'state',
         'users'
     ]
-    if rules is None:
-        single_rule = dict()
-        for key in rule_keys:
-            single_rule[key] = module.params[key]
-        rules = [single_rule]
+    new_rules = []
```
Comment (Andersson007) [inline plugins/modules/postgresql_pg_hba.py:1333]:
```suggestion
    new_rules = []
```
could we move the below lines to a separate function for logic readability?

**Review by Andersson007 — COMMENTED** [review]
@toydarian epic work, thanks a lot! My eyes feel happy to see the coverage!
From my limited perspective, there are two point:
- Proper parameter deprecation (see the comment), we should also create an issue for milestone 4.0.0 or 5.0.0
- And moving some code out of main to separate functions to improve general module logic understanding (see the comment, but it's up to you, just a suggestion)

File: plugins/modules/postgresql_pg_hba.py (hunk 1 above)
Comment (toydarian) [reply plugins/modules/postgresql_pg_hba.py:1281]: This actually doesn't do anything anymore. The module behaves as if this is always `true` (which doesn't break anything). So this is only here to ensure the module doesn't throw an error when this is specified. I added the additional fields.

File: plugins/modules/postgresql_pg_hba.py (hunk 3 above)
Comment (toydarian) [reply plugins/modules/postgresql_pg_hba.py:1336]: yes, it does, but I would do this in a later PR

File: plugins/modules/postgresql_pg_hba.py (hunk 2 above)
Comment (toydarian) [reply plugins/modules/postgresql_pg_hba.py:1375]: yes, it does, but I would do this in a later PR

File: plugins/modules/postgresql_pg_hba.py (hunk 1 above)
Comment (Andersson007) [reply plugins/modules/postgresql_pg_hba.py:1281]: @toydarian ok, thanks for explaining, I'll create an issue to properly deprecate it in another PR and remove it as we shouldn't keep any arguments that do nothing

**Review by Andersson007 — COMMENTED** [review]
@toydarian thanks for the change and comments!
One thing related to the latest changes: could you please add a line under `breaking_changes:` to the fragment announcing that the `keep_comments_at_rules` gonna get removed in 5.0.0?

**Comment by Andersson007** [comment]
@toydarian thanks for the contribution!
@hunleyd thanks for reviewing!
